internal/oidc/token/token_handler.go

```diff
@@ -165,9 +212,46 @@ func upstreamOIDCRefresh(ctx context.Context, session *psession.PinnipedSession,
 	return nil
 }
 
-func validateIdentityUnchangedSinceInitialLogin(validatedTokens *oidctypes.Token, session *psession.PinnipedSession, usernameClaimName string) error {
+func performUpstreamOIDCRefreshWithRetriesOnError(
+	ctx context.Context,
+	p provider.UpstreamOIDCIdentityProviderI,
+	s *psession.CustomSessionData,
+	retryOnErrorFactor float64,
+) (*oauth2.Token, error) {
+	var tokens *oauth2.Token
+
+	// For the default retryOnErrorFactor of 4.0 this backoff means...
+	// Try once, then retry upon error after sleeps of 50ms, 0.2s, 0.8s, 3.2s, and 12.8s.
+	// Give up after a total of 6 tries over ~17s if they all resulted in errors.
+	backoff := wait.Backoff{Steps: 6, Duration: 50 * time.Millisecond, Factor: retryOnErrorFactor}
+
+	isRetryableError := func(err error) bool {
+		plog.DebugErr("upstream refresh request failed in retry loop", err,
+			"providerName", s.ProviderName, "providerType", s.ProviderType, "providerUID", s.ProviderUID)
+		if ctx.Err() != nil {
+			return false // Stop retrying if the context was closed (cancelled or timed out).
+		}
+		retrieveError := &oauth2.RetrieveError{}
+		if errors.As(err, &retrieveError) {
+			return retrieveError.Response.StatusCode >= 500 // 5xx statuses are inconclusive and might be worth retrying.
+		}
+		return true // Retry any other errors, e.g. connection errors.
+	}
+
+	performRefreshOnce := func() error {
+		var err error
+		tokens, err = p.PerformRefresh(ctx, s.OIDC.UpstreamRefreshToken)
```

Comment:
We probably want to limit the context passed into this function such that each iteration can only consume so much of the over context timeout.
```suggestion
		n, cancel := context.WithTimeout(ctx, ...)
		defer cancel()
		tokens, err = p.PerformRefresh(n, s.OIDC.UpstreamRefreshToken)
```

Reply:
@enj Could you help me understand your suggestion better?

This context is the incoming http Request's context, passed through unchanged. I'm not aware of any timeouts that might be set onto it. It is passed through to handle potential cancellation of the context, which as you know can happen for incoming server requests for various reasons (client closes connection, etc).

Reply:
> I'm not aware of any timeouts that might be set onto it.

There should be some defaults configured on our server, but even if there are not, the client can choose whatever timeout it wants for the request.  We have to assume that we do not have infinite time.

> It is passed through to handle potential cancellation of the context, which as you know can happen for incoming server requests for various reasons (client closes connection, etc).

Consider the scenario:

1. Client has a 30 second timeout
2. The OIDC provider has two replicas, one of which is flakey

We attempt to perform an upstream refresh, and we hit the flakey replica.  Instead of the request failing with an error, it hangs (say due to connection issues).  Because this connection is hung, the first call to `tokens, err = p.PerformRefresh(ctx, s.OIDC.UpstreamRefreshToken)` takes up the entire context (because the client closes the connection in 30 seconds).  Thus we fail to retry because we let out first attempt use up all of our allocated time.

If we instead have:

```go
shortCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
defer cancel()
tokens, err = p.PerformRefresh(shortCtx, s.OIDC.UpstreamRefreshToken)
```

This way, each iteration of the retry loop (i.e. each call to `performRefreshOnce`) will be limited in how much overall time it will take.  Thus we have a stronger guarantee that we will actually retry.

Reply:
Oh I see what you mean now. Thanks.

I'm not sure what would be a good number to choose for a timeout. 10 seconds is a pretty short timeout because there could be a temporary network or compute slowness which could reasonably make an occasional request succeed after more than 10 seconds. Maybe 30 seconds? It's hard to put a specific number on it because whatever number we pick becomes a hard requirement that refresh requests to all OIDC IDPs on all varieties of networks must be able to succeed within that requirement or else users will occasionally lose their sessions.

Reply:
30 seconds sounds fine to me.  We allow one minute for discovery and login so presumably we need to have something less than that if we actually want to guarantee a retry:

https://github.com/vmware-tanzu/pinniped/blob/6c923d3bc6a516936589f088a2a5ead66b12c4ee/internal/controller/supervisorconfig/oidcupstreamwatcher/oidc_upstream_watcher.go#L454

https://github.com/vmware-tanzu/pinniped/blob/6c923d3bc6a516936589f088a2a5ead66b12c4ee/pkg/oidcclient/login.go#L303

https://github.com/vmware-tanzu/pinniped/blob/6c923d3bc6a516936589f088a2a5ead66b12c4ee/pkg/oidcclient/login.go#L445

Reply:
We could technically do something like 50 seconds to guarantee two retries while still giving the first attempt almost a minute.

Reply:
I pushed a new commit implementing this. Unfortunately this means that we lose test unit coverage on passing the
request's context through to the mocked PerformRefresh function. We used to be able to assert equality on the request's context because it was passed through unchanged. I don't know if there's a trick to be able to compare the timeout context.

Reply:
I am not too stressed by this, if you were really worried you could try to assert the `.Deadline()` on the context was within some range.  You could theoretically do a full comparison by creating an equivalent context based on the original context and the deadline, but I do not think that is worth the effort (and you would still need the deadline range check).